lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs

```diff
@@ -1145,6 +1176,195 @@ makeChangeForCoin targets excess =
+    => TokenQuantity
+    -- ^ The token quantity to be partitioned.
+    -> NonEmpty a
+    -- ^ Represents the number of portions in which to partition the quantity.
+    -> NonEmpty TokenQuantity
+    -- ^ The partitioned quantities.
+equipartitionTokenQuantity q =
+    fmap TokenQuantity . equipartitionNatural (unTokenQuantity q)
+
+--------------------------------------------------------------------------------
+-- Equipartitioning according to a maximum token quantity
+--------------------------------------------------------------------------------
+
+-- | Computes the equipartition of a token bundle into 'n' smaller bundles,
+--   according to the given maximum token quantity.
+--
+-- The value 'n' is computed automatically, and is the minimum value required
+-- to achieve the goal that no token quantity in any of the resulting bundles
+-- exceeds the maximum allowable token quantity.
+--
+equipartitionTokenBundleWithMaxQuantity
+    :: TokenBundle
+    -> TokenQuantity
+    -- ^ Maximum allowable token quantity.
+    -> NonEmpty TokenBundle
+    -- ^ The partitioned bundles.
+equipartitionTokenBundleWithMaxQuantity b maxQuantity =
+    NE.zipWith TokenBundle cs ms
+  where
+    cs = equipartitionCoin (view #coin b) ms
+    ms = equipartitionTokenMapWithMaxQuantity (view #tokens b) maxQuantity
+
+-- | Applies 'equipartitionTokenBundleWithMaxQuantity' to a list of bundles.
+--
+-- Only token bundles containing quantities that exceed the maximum token
+-- quantity will be partitioned.
+--
+-- If none of the bundles in the given list contain a quantity that exceeds
+-- the maximum token quantity, this function will return the original list.
+--
+equipartitionTokenBundlesWithMaxQuantity
+    :: NonEmpty TokenBundle
+    -- ^ Token bundles.
+    -> TokenQuantity
+    -- ^ Maximum allowable token quantity.
+    -> NonEmpty TokenBundle
+    -- ^ The partitioned bundles.
+equipartitionTokenBundlesWithMaxQuantity bs maxQuantity =
+    (`equipartitionTokenBundleWithMaxQuantity` maxQuantity) =<< bs
+
+-- | Computes the equipartition of a token map into 'n' smaller maps, according
+--   to the given maximum token quantity.
+--
+-- The value 'n' is computed automatically, and is the minimum value required
+-- to achieve the goal that no token quantity in any of the resulting maps
+-- exceeds the maximum allowable token quantity.
+--
+equipartitionTokenMapWithMaxQuantity
+    :: TokenMap
+    -> TokenQuantity
+    -- ^ Maximum allowable token quantity.
+    -> NonEmpty TokenMap
+    -- ^ The partitioned maps.
+equipartitionTokenMapWithMaxQuantity m (TokenQuantity maxQuantity)
+    | maxQuantity == 0 =
+        maxQuantityZeroError
+    | currentMaxQuantity <= maxQuantity =
+        m :| []
+    | otherwise =
+        equipartitionTokenMap m (() :| replicate extraPartCount ())
+  where
+    TokenQuantity currentMaxQuantity = TokenMap.maximumQuantity m
+
+    extraPartCount :: Int
+    extraPartCount = floor $ pred currentMaxQuantity % maxQuantity
```

Comment:
That's a slightly odd way to compute this and quite confusing at first. But ok with the construction of the non-empty list above :heavy_check_mark: 

That'd be clearer with a let-binding because this calculation only makes sense with the surrounding context :smirk:

Reply:
> That's a slightly odd way to compute this and quite confusing at first. But ok with the construction of the non-empty list above heavy_check_mark

Yes, it does seem odd at first. Though you're absolutely right -- it's this way because of the non-empty list.

The test case for this uses a more straightforward calculation:

```hs
-- | Computes the number of parts that 'equipartitionTokenMapWithMaxQuantity'
--   should return.
--
equipartitionTokenMapWithMaxQuantity_expectedLength
    :: TokenMap -> TokenQuantity -> Int
equipartitionTokenMapWithMaxQuantity_expectedLength
    m (TokenQuantity maxQuantity) =
        max 1 $ ceiling $ currentMaxQuantity % maxQuantity
  where
    TokenQuantity currentMaxQuantity = TokenMap.maximumQuantity m
```